ComSemApp/student/views.py

```diff
@@ -789,16 +790,46 @@ def get_object(self):
+        attempt_ids : list[str] = dict(self.request.GET)['attempt']
+        attempts : QuerySet[SpeakingPracticeAttempt] = SpeakingPracticeAttempt.objects.filter(student=self.student, pk__in=attempt_ids)
+        expressions : set[Expression] = set([attempt.expression for attempt in attempts])
+        familiarity_scores : dict[int, int] = {}
+
+        # In order to generate familiarity scores, we must get some data for all expressions.
+        worksheets : QuerySet[Worksheet] = self.course.worksheets.filter(status=teacher_constants.WORKSHEET_STATUS_RELEASED)
+        practice_attempts : QuerySet[SpeakingPracticeAttempt] = SpeakingPracticeAttempt.objects.filter(student=self.student, expression__worksheet__course=self.course)
+        completed : list[Worksheet] = []
+        exp_data : dict[Expression, dict[str, float | int]] = {}
+        WEIGHTS : dict[str,float] = {'correct_attempts': 0.15, 'days_since_review' : 0.1,  'wpm' : 0.25, 'last_score' : 0.5}
+        DATA_KEYS : list[str] = ['correct_attempts', 'days_since_review', 'wpm', 'last_score']
+
+        completed = [i for i in worksheets if i.complete_submission(self.student)]
+        for worksheet in completed:
+            expression_list : QuerySet[Expression] = worksheet.expressions.all().filter(Q(student=self.student) | Q(student=None) | Q(all_do=True))
+            for expression in expression_list:
+                exp_data[expression] = self.get_expression_data(practice_attempts.filter(expression=expression))
         
-        context = super(SpeakingPracticeResultsView, self).get_context_data(**kwargs)
-        context['practice_data'] = practice_data
+        mins = { key : min([y[key] for y in exp_data.values()]) for key in DATA_KEYS }
+        maxes = { key: max([y[key] for y in exp_data.values()]) for key in DATA_KEYS }
+        ranges = { key : maxes[key] - mins[key] if maxes[key] > mins[key] else 1 for key in DATA_KEYS }
+
+        for expression in expressions:
+            normalized_data : dict[str, int | float] = {}
+            normalized_data['correct_attempts'] = max(0, exp_data[expression]['correct_attempts'] - mins['correct_attempts']) / ranges['correct_attempts']
+            normalized_data['days_since_review'] = 1 - (max(0, exp_data[expression]['days_since_review'] - mins['days_since_review']) / ranges['days_since_review'])
+            normalized_data['wpm'] = max(0, exp_data[expression]['wpm'] - mins['wpm']) / ranges['wpm']
+            normalized_data['last_score'] = max(0, exp_data[expression]['last_score'] - mins['last_score']) / ranges['last_score']
+            familiarity_scores[expression.pk] = round(sum([float(normalized_data[key]) * WEIGHTS[key] for key in DATA_KEYS]) * 100)
+
+        context['attempts'] = [{'id' : attempt.pk, 'transcription' : attempt.transcription, 'audio_path' : attempt.audio.path, 'familiarity_score' : familiarity_scores[attempt.expression.pk]} for attempt in attempts]
```

Comment:
Similar to my confusion above, shouldn't the results context data be passing things like the correct expression rather than familiarity score?

Reply:
I can add correct expression. I believe the familiarity score was being passed so we can display it to the user. It can be removed